stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,780 @@
+                    sig_share_response.sign_id, sig_share_response.signer_id, self.ids_to_await
+                );
+            }
+            _ => {}
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate signature
+            let polys: Vec<PolyCommitment> = self
+                .dkg_public_shares
+                .values()
+                .map(|ps| ps.public_share.clone())
+                .collect();
+
+            let nonce_responses = (0..self.total_signers).map(|i| self.public_nonces[&i].clone()).collect::<Vec<NonceResponse>>();
+
+            let nonces = nonce_responses
+                .iter()
+                .flat_map(|nr| nr.nonces.clone())
+                .collect::<Vec<PublicNonce>>();
+
+            let shares = &self
+                .public_nonces
+                .iter()
+                .flat_map(|(i, _)| self.signature_shares[i].clone())
+                .collect::<Vec<SignatureShare>>();
+
+            info!(
+                "aggregator.sign({:?}, {:?}, {:?})",
+                self.message,
+                nonces.len(),
+                shares.len()
+            );
+
+            let mut aggregator = v1::SignatureAggregator::new(self.total_keys, self.threshold, polys)?;
+
+            let sig = aggregator.sign(&self.message, &nonces, shares)?;
+
+            info!("Signature ({}, {})", sig.R, sig.z);
+
+            let proof = SchnorrProof::new(&sig)?;
+
+            info!("SchnorrProof ({}, {})", proof.r, proof.s);
+
+            if !proof.verify(&self.aggregate_public_key.x(), &self.message) {
+                warn!("SchnorrProof failed to verify!");
+                return Err(Error::SchnorrProofFailed);
+            }
+
+            self.move_to(State::Idle)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, PartialEq)]
+/// Coordinator states
+pub enum State {
+    /// The coordinator is idle
+    Idle,
+    /// The coordinator is distributing public shares
+    DkgPublicDistribute,
+    /// The coordinator is gathering public shares
+    DkgPublicGather,
+    /// The coordinator is distributing private shares
+    DkgPrivateDistribute,
+    /// The coordinator is gathering DKG End messages
+    DkgEndGather,
+    /// The coordinator is requesting nonces
+    NonceRequest,
+    /// The coordinator is gathering nonces
+    NonceGather,
+    /// The coordinator is requesting signature shares
+    SigShareRequest,
+    /// The coordinator is gathering signature shares
+    SigShareGather,
+}
+
+/// The state machine trait for the frost coordinator
+pub trait StateMachine {
+    /// Attempt to move the state machine to a new state
+    fn move_to(&mut self, state: State) -> Result<(), Error>;
+    /// Check if the state machine can move to a new state
+    fn can_move_to(&self, state: &State) -> Result<(), Error>;
+}
+
+impl StateMachine for Coordinator {
+    fn move_to(&mut self, state: State) -> Result<(), Error> {
+        self.can_move_to(&state)?;
```

Comment:
I love how this system checks for a valid state transition!  Do you think it makes sense to outright `panic!()` if we encounter an invalid state transaction? Because, that would indicate a logic bug in the signer, right?

Reply:
I have a moderate distaste for `panic!` in anything that runs in a network context.  Having said that, we `panic!` if we fail to sign messages, since that means our private key is completely broken and we can't reasonably continue.  

For state machine errors, I can see a possibility of getting an unexpected message that triggers a bad state machine change.  If we can handle that kind of thing without `panic!` I think we should.